LoRaWAN: Retiring LoRaWANBase class

hasnainvirk opened the pull request:
### Description

It was decided to retire LoRaWANBase class which served
as a pure virtual interface class from which LoRaWAN network stack
implementations would get inherited. However, the current view is that
we may be the only user of it so we could retire LoRaWANBase.


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [X] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->

@AnttiKauppila @kjbracey-arm 

### Release Notes
There has been only one user of the LoRaWANBase class, the LoRaWANInterface. We anticipated that the third parties may use this class to provide their own LoRaWAN network stacks alongwith the default network stack provided by Mbed OS. However, a trend has been observed that the default stack provided by Mbed OS has become the de-facto choice. Having LoRaWANBase class adds 1K to the ROM in terms of footprint (for GCC, ARMCC and IAR optimize that already). It has been decided to do away with the LoRaWANBase class for now, and if the need arises it can be reintroduced back.

Commit:
Retiring LoRaWANBase class

It was decided within the team to retire LoRaWANBase class which served
as a pure virtual interface class from which LoRaWAN network stack
implementations would get inherited. However, the current view is that
we may be the only user of it so we could retire LoRaWANBase.

ciarmcom commented:
@hasnainvirk, thank you for your changes.
@AnttiKauppila @kjbracey-arm@ARMmbed/mbed-os-wan @ARMmbed/mbed-os-maintainers please review.

Commit:
Deprecation notice for LoRaWANBase

A deprecation notice has been added for any users of LoRaWANBase and any
existing reference is redirected to LoRaWANInterface.

mattbrown015 commented:
Hi @hasnainvirk,

I like this PR! :-)

In our build we modify `LoRaWANInterface` so that it no longer inherits from `LoRaWANBase`.

But, crucially, we also removed all the 'virtual' keywords from the member functions.

Making the member functions non-virtual reduced our executable size by over a 1000 bytes.  1000 bytes is a significant saving.  My feeling is that making the functions non-virtual meant the linker could see that they weren't used and remove them from the executable.  Obviously other people's mileage would vary.

Do you believe that `LoRaWANInterface` needs to be polymorphic?

hasnainvirk commented:
@mattbrown015 Unfortunately we need to pay the price of 1000 bytes here because we wish to leave a window for application developers to write a single application that can be used with any LoRaWAN network stack. Actually that was the sole purpose of the dear departed LoRaWANBase :).
Now that she's gone, we still need the APIs in LoRaWANInterface to be overridable.  I wouldn't mind having a 1000 bytes back, but this is a thing which we need to discuss internally first as it may  not be entirely in line with Mbed philosophy.

mattbrown015 commented:
It's fine by me if the functions remain virtual.  Our build process has a mechanism for applying patches to make changes that we want and while not perfect it suits our needs.

But, I'm interested because I'm interested in C++ and API design! :-)

Is making the class _run-time polymorphic_ the only way to make it flexible?

If an alternative implementation is required is it not possible to just override non-virtual functions in a new derived class?

Virtual functions are only necessary if the app needs to pass around base class pointers without caring which stack implementation is being used.  Or to put it another way, do you expect clients to select the stack implementation at run-time.

If we were working on server software running on x64s we would probably put on our best C++ hats and  program everything to interfaces.  But we're embedded so we probably want to leverage compile-time techniques.

(Feel free not to reply, both of us should probably be concentrating on real work!)

cmonr commented:
> But, I'm interested because I'm interested in C++ and API design! :-)

Ping @kjbracey-arm @pan- ;)

mattbrown015 commented:
I thought of another way of looking at it!

The phy class is polymorphic.

Currently we statically construct a EU868 phy and pass it to LoRaWAN as a base class pointer.  In this case the polymorphism is an overhead that we're not using.  But, I can see that a mobile device might need to change its phy at run-time as it moves RF jurisdictions.  Or, perhaps more likely, production is simplified by  building all devices to the same spec and then configuring them for the intended jurisdiction when they are deployed.

Comparing the phy to stack implementation.  The API currently adds overhead for all your customers on the basis that you think you might have a few customers that want to select the stack implementation at run-time.

It's your API.  You need to decide if the price paid by all your customers is worth it to make life easier for a few.

kjbracey commented:
@mattbrown015  I'm curious - was your 1000 byte figure with GCC or one of the other toolchains? GCC is particularly disadvantaged here, as it can't do unused virtual function elimination, unlike IAR and ARM C.

We certainly want/need the polymorphism for the cases where we are passing around stuff where there are multiple viable implementations `NetworkInterface` and `Socket` in particular.

I would like to regard optimising it as a tools issue - they should be capable of eliminating unused virtual stuff generally, and with Link-Time Optimisation, it should even be possible to totally eliminate the virtuals if the compiler+linker can see that only one derived type is instantiated. 

Maybe I'm unduly optimistic though.

mattbrown015 commented:
Hi @kjbracey-arm, yes we're using ARM GCC 7.  I've just noticed that ARM GCC 8 was released in December; that means Mbed OS is even further behind in its support for ARM GCC 6. ;-)

It's interesting that this is an example of where the more expensive compilers add some value.  We won't be changing compiler for our project though, in theory we're nearly finished!

If you have concrete examples that justify your design that's fair enough, like I say, it's your API! :-)

pan- commented:
> Maybe I'm unduly optimistic though.

I think it would be interesting to do some tests with Link Time Optimization (LTO) enabled to identify and document the pattern that can be optimised and those that can't. The issue lies in making sure that the linker optimise the application as a whole and do not consider dynamic link which prevent vtable deletion.

hasnainvirk commented:
@kjbracey-arm @pan- @AnttiKauppila I suggest that we do away with virtual modifier in LoRaWANInterface as it brings little value after retiring LoRaWANBase. 
If not, then we shouldn't retire LoRaWANBase at all. Because if a third party wishes to provide an implementation of their stack it would make sense to inherit from LoRaWANBase. LoRaWANInterface have embedded objects of LoRaWANStack class and other system level classes which the third party may not require.

Commit:
Removing virtual modifier from LoRaWANInterface

Doing away with virtual modifier from LoRaWANInterface, gets rid of
vtable for LoRaWANInterface.

AnttiKauppila commented:
@hasnainvirk You are correct. LoRaWANStack is a bit problematic currently. But only because bind_phy_and_radio_driver() is called in the constructor (Cannot be overwritten). We need to think more about this. If we init the radio in constructor and call bind later, it should work.

Commit:
Doxygen corrections

Adding group identidier so that LoRaWANInterface class goes to the class
hierarchy section rather than data-structures.

Adding missing documentation for a couple of public functions.

Adding \code and \endcode modifiers for the example code in the
documentation.

Adding compile time NO_DOXYGEN flag for the implementations of the
LoRaPHY Class.

Adding documentation for some of the private structures.

hasnainvirk commented:
@AnttiKauppila In https://github.com/ARMmbed/mbed-os/pull/9219/commits/d05e7becf81e3317316f105bc09304ed4a797710 we have done away with virtual modifier. In the near future we can go forward for more refactoring involving LoRaWANStack, maybe like combining LoRaWANInterface and LoRaWANStack.

AnttiKauppila reviewed (changes requested):
I want to have a blessing from architects to get rid of virtual methods. It has been Mbed OS design to be able to override connectivity stacks

hasnainvirk commented:
@AnttiKauppila Totally agree. We need to know if that is alright with them. @kjbracey-arm  Any comments ?

kjbracey commented:
I don't have any strong feelings. The moment you do have a second stack, you can reintroduce the base class, or have a virtual implementation class. For now, it's not buying you anything.

If it's just an issue of optimisation, in C++11, I think marking LoRAWANinterface `final` might reclaim all the vtable overhead. Might be worth trying.

pan- commented:
@kjbracey-arm I've made some tests on ARM GCC8 with c++14 enabled as well as `lto` and `-fipa-pta` optimisations turned on. Turns out that GCC is able to devirtualize calls however it does not remove unused functions from the vtable which continue to be assigned to polymorphic classes. 

Cost of unused virtual functions is an issue for BLE. I'm exploring other options such as static polymorphism via CRTP which might be viable when there is only one implementation selected at compile time or when instances are not accessed through their base class.

mattbrown015 commented:
CRTP, interesting! :-)

That's exactly the sort of idea I was fishing for when I said "Is making the class run-time polymorphic the only way to make it flexible?"

While aware of CRTP it's not something I've ever used.  I'm going to read about it while I eat my sandwiches!

hasnainvirk commented:
@0xc0170 This can be merged now.

0xc0170 commented:
This deprecation would be worth having in the release notes. Shall we add release notes section (please add new section following https://github.com/ARMmbed/mbed-os-5-docs/pull/933 - not yet published).

@AnotherButler Please review docs changes in this PR plus also the release notes section once updated

hasnainvirk commented:
@0xc0170 @AnotherButler Release notes section added.

0xc0170 commented:
@donatieng please review (for @bulislaw)

hasnainvirk commented:
@0xc0170 Can we please prioritize that for CI and merge ?

0xc0170 commented:
> @0xc0170 Can we please prioritize that for CI and merge ?

I would like to have this reviewed by @donatieng or @bulislaw . IN the meantime we can start CI

0xc0170 commented:
CI started

ashok-rao reviewed (approved):
Approved based on email discussions with Etteplan. Thanks!
